File: programmers/공원 산책.js
function solution(park, routes) {
  const a = park.length;
  const b = park[0].length;

  // 시작 좌표 설정
  let start;
  for (let i = 0; i < a; i++) {
    for (let j = 0; j < b; j++) {
      if (park[i][j] == "S") start = [i, j];
    }
  }

  // 방향이동
  const directions = {
    E: [0, 1],
    W: [0, -1],
    S: [1, 0],
    N: [-1, 0],
  };

  let ans = [0, 0];
  // 주어진 이동정보 배열에 담기
  for (const route of routes) {
    const [dir, disStr] = route.split(" ");
    let distance = parseInt(disStr);
    let [nx, ny] = start;
    // 주어진 걸음수 만큼 한칸씩 이동
    let step = 0;
    while (step < distance) {
      nx += directions[dir][0];
      ny += directions[dir][1];
      // 만약 밖에 나가게 되거나 X를 만나게 된다면 종료
      if (nx < 0 || a <= nx || ny < 0 || b <= ny || park[nx][ny] === "X") break;
      step++;
    }
    // 원하는 걸음수를 채우면 start는 마지막 위치로 바꿈
    if (step === distance) start = [nx, ny];
  }

  return start;
}

File: Java/5월/05_28/공원산책.java
import java.util.*;

class Solution {
    public int[] solution(String[] park, String[] routes) {

        // 2차원 배열 저장
        char[][] map = new char[park.length][park[0].length()];

        for (int i = 0; i < park.length; i++) {
            for (int j = 0; j < park[i].length(); j++) {
                map[i][j] = park[i].charAt(j);
            }
        }

        int x = 0;
        int y = 0;

        // 시작 위치 S 찾기
        for (int i = 0; i < park.length; i++) {
            for (int j = 0; j < park[i].length(); j++) {

                if (map[i][j] == 'S') {
                    y = i;
                    x = j;
                }
            }
        }

        // 동 서 남 북
        char[] dirArr = {'E', 'W', 'S', 'N'};

        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};

        // 이동
        for (String route : routes) {

            String[] parts = route.split(" ");

            char dir = parts[0].charAt(0);
            int distance = Integer.parseInt(parts[1]);

            int nx = x;
            int ny = y;

            // 방향 찾기
            for (int i = 0; i < dirArr.length; i++) {

                if (dir == dirArr[i]) {

                    boolean possible = true;

                    // 한 칸씩 이동 검사
                    for (int d = 1; d <= distance; d++) {

                        int tx = x + dx[i] * d;
                        int ty = y + dy[i] * d;

                        // 범위 밖
                        if (tx < 0 || ty < 0 ||
                            ty >= park.length ||
                            tx >= park[0].length()) {

                            possible = false;
                            break;
                        }

                        // 장애물
                        if (map[ty][tx] == 'X') {
                            possible = false;
                            break;
                        }

                        nx = tx;
                        ny = ty;
                    }

                    // 이동 가능하면 적용
                    if (possible) {
                        x = nx;
                        y = ny;
                    }

                    break;
                }
            }
        }

        return new int[]{y, x};
    }
}